lib/ansible_tower_client/base_model.rb

```diff
@@ -1,9 +1,152 @@
+
+      # Find the exclusion list for the model of next level (@embed_model)
+      # '#' is the separator between levels. Remove attributes
+      # before the first separator.
+      @child_excl_list = self.class.send(:excl_list).map do |e|
+        e.index('#') ? e[e.index('#') + 1..-1] : ''
+      end
+
+      __setobj__(json.kind_of?(Hash) ? json.dup : JSON.parse(json))
+    end
+
+    def to_h
+      __getobj__.dup
+    end
+
+    def to_hash
+      __getobj__.dup
+    end
+
+    def to_json
+      __getobj__.to_json
+    end
+
+    def to_s
+      __getobj__.to_json
+    end
+
+    def inspect
+      string = "<#{self.class} "
+      method_list = methods(false).select { |m| !m.to_s.include?('=') }
+      string << method_list.map { |m| "#{m}=#{send(m).inspect}" }.join(", ")
+      string << ">"
+    end
+
+    def ==(other)
+      return false unless other.kind_of?(BaseModel)
+      __getobj__ == other.__getobj__
+    end
+
+    def eql?(other)
+      return false unless other.kind_of?(BaseModel)
+      __getobj__.eql?(other.__getobj__)
+    end
+
+    # Support hash style accessors
+    def [](key)
+      __getobj__[key]
+    end
+
+    def []=(key, val)
+      key_exists = __getobj__.include?(key)
+      __getobj__[key] = val
+
+      return if key_exists
+      add_accessor_methods(snake_case(key), key)
+    end
+
+    protected
+
+    def __getobj__
+      @hashobj
+    end
+
+    # Create snake_case accessor methods for all hash attributes
+    # Use _alias if an accessor conflicts with existing methods
+    def __setobj__(obj)
+      @hashobj = obj
+      excl_list = self.class.send(:excl_list)
+      idkeys = nil
+      obj.each do |key, value|
+        snake = snake_case(key)
+        unless excl_list.include?(snake) # Must deal with nested models
+          if value.kind_of?(Array)
+            newval = value.map { |elem| elem.kind_of?(Hash) ? nested_object(snake.camelize.singularize, elem) : elem }
+            obj[key] = newval
+          elsif value.kind_of?(Hash)
+            obj[key] = nested_object(snake.camelize, value)
+
+            # will treat attributes that appear in 'related' section as ids.
+            idkeys = value.keys if key == 'related'
+          end
+        end
+
+        snake = "#{snake}_id" if idkeys && !snake.end_with?('_id') && idkeys.include?(key)
+        add_accessor_methods(snake, key)
+      end
+    end
+
+    def nested_object(klass_name, value)
+      unless self.class.const_defined?(klass_name)
+        child_excl_list = @child_excl_list
+        self.class.const_set(klass_name, Class.new(BaseModel) { attr_hash(*child_excl_list) })
+      end
+      self.class.const_get(klass_name).new(api, value)
+    end
+
+    def add_accessor_methods(method, key)
+      method.prepend('_') if methods.include?(method.to_sym)
+      instance_eval { define_singleton_method(method) { __getobj__[key] } }
+      instance_eval { define_singleton_method("#{method}=") { |val| __getobj__[key] = val } }
```

Comment:
The model class when used in armrest gem does not contain the service class/provider connection. It services purely as an attributes container. Therefore it makes sense to support both read and write.

In this gem, if you expect any changes in the model to be in sync with its corresponding provider object, then we should not auto_generate the setter.